File: 3171-find-subarray-with-bitwise-and-closest-to-k/3171-find-subarray-with-bitwise-and-closest-to-k.java
class Solution {
    public int minimumDifference(int[] nums, int k) {
        Set<Integer>prev = new HashSet<>();
        int ans = Integer.MAX_VALUE;
        for(int x:nums) {
            Set<Integer>next = new HashSet<>();
            for(int p:prev) {
                next.add(p&x);
            }
            next.add(x);
            for(int n:next) {
                ans = Math.min(ans, Math.abs(n - k));
            }
            prev = next;
        }
        return ans;
    }
}

File: 3171-find-subarray-with-bitwise-and-closest-to-k/3171-find-subarray-with-bitwise-and-closest-to-k.cpp
class Solution {
public:
    int minimumDifference(vector<int>& nums, int k) {
        set<int>prev;
        int ans = INT_MAX;
        for(int x:nums) {
            set<int>next;
            for(int p:prev) {
                next.insert(x&p);
            }
            next.insert(x);
            for(int n:next) {
                ans = min(ans, abs(n - k));
            }
            prev = next;
        }
        return ans;
    }
};